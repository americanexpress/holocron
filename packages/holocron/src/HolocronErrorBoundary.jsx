import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';

class HolocronErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { error: null };
  }

  static getDerivedStateFromError(error) {
    return { error };
  }

  // eslint-disable-next-line class-methods-use-this
  componentDidCatch(error, errorInfo) {
    console.error(error);
    console.error(errorInfo);
  }

  render() {
    const { props: { children }, state: { error } } = this;

    if (error) {
      if (process.env.NODE_ENV === 'production') return null;
      return <div style={{ color: 'red' }}>{error.message}</div>;
    }

    return (
      <Fragment>
        {children}
      </Fragment>
    );
  }
}

HolocronErrorBoundary.propTypes = {
  children: PropTypes.node,
};

HolocronErrorBoundary.defaultProps = {
  children: null,
};

export const withHolocronErrorBoundary = (WrappedComponent) => {
  const name = WrappedComponent.displayName || WrappedComponent.name;
  const Wrapped = props => (
    <HolocronErrorBoundary>
      <WrappedComponent {...props} />
    </HolocronErrorBoundary>
  );

  Wrapped.displayName = name
    ? `WithHolocronErrorBoundary(${name})`
    : 'WithHolocronErrorBoundary';

  return hoistNonReactStatics(Wrapped, WrappedComponent);
};

export default HolocronErrorBoundary;
