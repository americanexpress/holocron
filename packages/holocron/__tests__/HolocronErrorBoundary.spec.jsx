import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { mount } from 'enzyme';
import HolocronErrorBoundary, { withHolocronErrorBoundary } from '../src/HolocronErrorBoundary';

describe('HolocronErrorBoundary', () => {
  let consoleErrorSpy;
  let mockError;
  let CharacterWelcome;
  let Character;

  const { NODE_ENV } = process.env;

  beforeAll(() => {
    consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    mockError = new Error(
      'These are not the droids you are looking for.'
    );

    // eslint-disable-next-line react/prop-types
    Character = ({ name }) => {
      switch (name) {
        case 'R2-D2':
        case 'C-3PO':
        case 'BB-8':
          throw mockError;

        default:
          return <p>{`Welcome ${name}`}</p>;
      }
    };

    // eslint-disable-next-line react/prop-types
    CharacterWelcome = ({ name }) => (
      <div>
        <p>{`Looking up character ${name}`}</p>
        <Character name={name} />
      </div>
    );
  });

  afterAll(() => {
    consoleErrorSpy.mockRestore();
  });

  afterEach(() => {
    process.env.NODE_ENV = NODE_ENV;
  });

  it('Renders the child component if there is no error', () => {
    const wrapper = mount(
      <HolocronErrorBoundary>
        <CharacterWelcome name="Chewbacca" />
      </HolocronErrorBoundary>
    );

    const CharacterWelcomeWithHolocronErrorBoundary = withHolocronErrorBoundary(CharacterWelcome);
    const wrapperWithHolocronErrorBoundary = mount(
      <CharacterWelcomeWithHolocronErrorBoundary name="Chewbacca" />
    );

    expect(wrapper.state().error).toBe(null);

    expect(
      wrapper.contains(<CharacterWelcome name="Chewbacca" />)
    ).toBe(true);

    expect(
      wrapperWithHolocronErrorBoundary.find(HolocronErrorBoundary).instance().state.error
    ).toBe(null);

    expect(
      wrapperWithHolocronErrorBoundary.contains(
        <CharacterWelcome name="Chewbacca" />
      )
    ).toBe(true);
  });

  it('Sets its state to an error state and renders the error message', () => {
    const wrapper = mount(
      <HolocronErrorBoundary>
        <CharacterWelcome name="R2-D2" />
      </HolocronErrorBoundary>
    );

    const CharacterWelcomeWithHolocronErrorBoundary = withHolocronErrorBoundary(CharacterWelcome);
    const wrapperWithHolocronErrorBoundary = mount(
      <CharacterWelcomeWithHolocronErrorBoundary name="R2-D2" />
    );

    expect(wrapper.state().error).toEqual(expect.objectContaining(mockError));
    expect(
      wrapperWithHolocronErrorBoundary.find(HolocronErrorBoundary).instance().state.error
    ).toEqual(expect.objectContaining(mockError));
    expect(
      wrapper.containsMatchingElement(
        <div>
          {mockError.message}
        </div>
      )
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('Sets its state to an error state and renders null in production', () => {
    process.env.NODE_ENV = 'production';
    const wrapper = mount(
      <HolocronErrorBoundary>
        <CharacterWelcome name="C-3PO" />
      </HolocronErrorBoundary>
    );

    const CharacterWelcomeWithHolocronErrorBoundary = withHolocronErrorBoundary(CharacterWelcome);
    const wrapperWithHolocronErrorBoundary = mount(
      <CharacterWelcomeWithHolocronErrorBoundary name="C-3PO" />
    );

    expect(wrapper.state().error).toEqual(expect.objectContaining(mockError));
    expect(
      wrapperWithHolocronErrorBoundary.find(HolocronErrorBoundary).instance().state.error
    ).toEqual(expect.objectContaining(mockError));
    expect(
      wrapper.containsMatchingElement(
        <HolocronErrorBoundary />
      )
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('Sets the correct displayName for the wrapped component', () => {
    function NormalModule() {
      return null;
    }

    expect(withHolocronErrorBoundary(NormalModule).displayName).toBe(
      'WithHolocronErrorBoundary(NormalModule)'
    );

    function ModuleWithDisplayNameOverride() {
      return null;
    }

    ModuleWithDisplayNameOverride.displayName = 'ModuleOverride';

    expect(
      withHolocronErrorBoundary(ModuleWithDisplayNameOverride).displayName
    ).toBe('WithHolocronErrorBoundary(ModuleOverride)');
    expect(withHolocronErrorBoundary(() => null).displayName).toBe('WithHolocronErrorBoundary');
  });
});
