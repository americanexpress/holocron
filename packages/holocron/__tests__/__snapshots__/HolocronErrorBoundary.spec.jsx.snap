// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HolocronErrorBoundary Sets its state to an error state and renders null in production 1`] = `<HolocronErrorBoundary />`;

exports[`HolocronErrorBoundary Sets its state to an error state and renders the error message 1`] = `
<HolocronErrorBoundary>
  <div
    style={
      Object {
        "color": "red",
      }
    }
  >
    These are not the droids you are looking for.
  </div>
</HolocronErrorBoundary>
`;
